File: 2022/05/solution.py
def parse_input():
    with open('input.txt') as f:
        input = f.read()
    (cargo, instructions) = [
        section.strip().split('\n')
        for section
        in input.split('\n\n')
    ]

    stack_count = (len(cargo[0]) + 1) // 4
    stacks = [[] for _ in range(stack_count)]
    for row in cargo[:-1]:
        for i in range(stack_count):
            crate = row[i * 4 + 1]
            if crate != ' ':
                stacks[i].append(crate)
    for stack in stacks:
        stack.reverse()

    moves = []
    for instruction in instructions:
        split = instruction.split()
        moves.append(tuple(map(int, (split[1], split[3], split[5]))))

    return stacks, moves


def part1(input):
    stacks = [stack.copy() for stack in input[0]]
    for n, source, destination in input[1]:
        for _ in range(n):
            stacks[destination - 1].append(stacks[source - 1].pop())
    return ''.join(stack[-1] for stack in stacks)


def part2(input):
    stacks = [stack.copy() for stack in input[0]]
    for n, source, destination in input[1]:
        stacks[destination - 1].extend(stacks[source - 1][-n:])
        stacks[source - 1] = stacks[source - 1][:-n]
    return ''.join(stack[-1] for stack in stacks)


input = parse_input()
print(part1(input))
print(part2(input))

File: 2022/05/solution.rs
use std::fs;

fn main() {
    let input = parse_input();
    println!("{}", part1(&input));
    println!("{}", part2(&input));
}

fn parse_input() -> (Vec<Vec<char>>, Vec<[usize; 3]>) {
    let input = fs::read_to_string("input.txt").unwrap();
    let (cargo, instructions) = input.split_once("\n\n").unwrap();

    let cargo_rows: Vec<_> = cargo.lines().collect();
    let stack_count = (cargo_rows[0].len() + 1) / 4;
    let mut stacks = vec![vec![]; stack_count];
    cargo_rows
        .iter()
        .take(cargo_rows.len() - 1)
        .for_each(|row| {
            row.chars()
                .skip(1)
                .step_by(4)
                .enumerate()
                .for_each(|(i, cargo_crate)| {
                    if cargo_crate != ' ' {
                        stacks[i].push(cargo_crate);
                    }
                });
        });
    stacks.iter_mut().for_each(|stack| stack.reverse());

    let moves: Vec<_> = instructions
        .lines()
        .map(|line| {
            let move_values: Vec<_> = line
                .split(' ')
                .filter_map(|x| x.parse::<usize>().ok())
                .collect();
            [move_values[0], move_values[1] - 1, move_values[2] - 1]
        })
        .collect();

    (stacks, moves)
}

fn part1(input: &(Vec<Vec<char>>, Vec<[usize; 3]>)) -> String {
    let mut stacks: Vec<_> = input.0.to_vec();
    input.1.iter().for_each(|&[n, source, destination]| {
        (0..n).for_each(|_| {
            let cargo_crate = stacks[source].pop().unwrap();
            stacks[destination].push(cargo_crate);
        })
    });
    stacks.iter().map(|stack| stack.last().unwrap()).collect()
}

fn part2(input: &(Vec<Vec<char>>, Vec<[usize; 3]>)) -> String {
    let mut stacks: Vec<_> = input.0.to_vec();
    input.1.iter().for_each(|&[n, source, destination]| {
        let range_start = stacks[source].len() - n;
        let cargo_crates: Vec<_> = stacks[source].drain(range_start..).collect();
        stacks[destination].extend(cargo_crates.iter());
    });
    stacks.iter().map(|stack| stack.last().unwrap()).collect()
}
